### ISSUE TITLE
Add optional batch normalization layer

### ISSUE THREAD

**Issue by bigtreeliujie** [reinforceio/tensorforce#135]
Batch normalization layer ([paper](https://arxiv.org/pdf/1502.03167.pdf)) is widely used when training deep networks. It appears that batch normalization make the network learning faster and generalization better. It is very easy to add batch normalization layers in core/networks/layers within tensorflow, as there is a implementation in tensorflow.contrib.layers.layer_norm.

**Comment by bigtreeliujie** [comment]
I have read your blog and know that  you leave things like batch normalization to  the user. So a close this issue.

**Comment by lefnire** [comment]
batch norm should be as simple as creating a custom layer as such:

```python
def batch_norm(x, size=None, scope=None, summary_level=0):
    with tf.variable_scope(scope):
        x = tf.contrib.layers.batch_norm(x, is_training=True) # or tf.nn.batch_normalization()?
    return x
```

but the problem is it requires an extra step adding `tf.GraphKeys.UPDATE_OPS` to the training step:

```python
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
    train_step = tf.train.GradientDescentOptimizer(0.01).minimize(loss)
sess = tf.Session()
...
```

(see [this blog post](http://ruishu.io/2016/12/27/batchnorm/)). So there might be some complication that would prevent users from handling this on their own at present, the way TensorForce abstracts session-runs. I haven't dug deep to find out, but a quick code-search for `UPDATE_OPS` in the repo shows not currently handled.

**Comment by michaelschaarschmidt** [comment]
I would add that batch norm is not widely used in RL which is why it was not a priority for us.

**Comment by lefnire** [comment]
Interesting! I implemented batch_norm for my project & it didn't help despite the researcher oohs and aahs, figured it didn't apply to my thing or was doing it wrong - glad to know. Incidentally feature scaling / norm actually _hurts_ performance for my project too, dunno if that's also an RL thing (I seem to recall hearing that somewhere). 

Yeah keep it open, I'll take a crack at it when I'm feeling motivated. Might be handy if nothing more than adding that `GraphKeys.UPDATE_OPS` bit for additional custom layer flexibility.

**Comment by michaelschaarschmidt** [comment]
You can also experiment with selu activations (self normalizing networks), but we didnt find much impact either..

**Comment by AlexKuhnle** [comment]
Side note: I think batch normalization is mostly popular for CNN vision stuff to get a few percent recognition accuracy more. In reinforcement learning it's probably more things like reward normalization that consistently improve results.

**Comment by Kismuz** [comment]
From my experience I can add that batch-norm is not well suited for some of RL algorithms, especially ones using context-aware policy estimators (LSTM). This is due to the fact such estimator processes batches of consecutive correlated experiences (which is trajectory in obs. space) and thus provides batch-norm layer with highly biased distribution statistics. 
Shortly, to learn proper "normalisation weights" batch-norm layer needs batches of shuffled experiences ( feedforward CNN etc.) 
Actually when using LSTM-based estimators (and no experience replay, like in vanilla A3C) batch-norm can significantly hurt performance.

**Comment by lefnire** [comment]
Hearing all this, I probably won't find the motivation to implement the feature ;) Up for grabs if someone else; else feel free to close

**Comment by AlexKuhnle** [comment]
Okay, we'll close it, if anyone wants to implement it, it's still basically described in the blog post.
